doc/reviewing-contributions: Add points about patches (#240136)
NixOS/nixpkgs: doc/contributing/reviewing-contributions.chapter.md

- [ ] package build on ARCHITECTURE
- [ ] executables tested on ARCHITECTURE
- [ ] all depending packages build
- [ ] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed

> - [ ] patches that are remotely available are fetched with `fetchpatch`

Reviewer: I would like to either not mention how patches should be fetched, or explain the nuances of when to use `fetchpatch` vs `fetchurl`, because I see a lot of misunderstanding about it.  (Maybe this would make more sense elsewhere in the manual, and we could link to it here.)

`fetchpatch` for when a patch is generated in response to the request for it — .patch URLs provided by GitHub, for example.  It normalises the patches so that it (hopefully) won't have a different hash the next time it's fetched, which is good, but that comes at the consequence of complexity.  One of the main downsides to `fetchpatch` is that it depends on a specific, old version of patchutils to do this normalization, and because we can't ever change that, we now have `fetchpatch2`, which uses a newer version.  Additionally, some packages that are early enough in the bootstrap chain can't use `fetchpatch`, because they are dependencies of patchutils.

So for patches that are static files (e.g. patches checked into another distro's git repo, or raw messages from a mailing list archive), it's preferable to use `fetchurl` so we aren't adding even more dependencies on the behaviour of a specific version of patchutils when we don't need to.

Author: Indeed this doesn't seem to have been documented yet. I've split that into a separate issue, #240141, because it's a related pre-existing problem.

- [ ] source is fetched using the appropriate function
- [ ] the list of `phases` is not overridden
- [ ] when a phase (like `installPhase`) is overridden it starts with `runHook preInstall` and ends with `runHook postInstall`.
- [ ] patches have a comment describing either the upstream URL or a reason why the patch wasn't upstreamed

> - [ ] patches that are remotely available are fetched with `fetchpatch`

Author: ```suggestion
- [ ] patches that are remotely available are fetched rather than vendored
```
